Added email field for user account

rashmimaharana opened the pull request:
Hi @drakkan 

I have added email field for user account and have done respective changes in sql. Please verify and let me know if you have any issue, i will do those changes.

Thanks,
Rashmi Maharana

drakkan commented:
Thanks but you should prepare a database migration, otherwise email will be available only for new installations and database queries fail if someone upgrade an old installation.

Email should also be validated as we do for admins otherwise we accept any string, please take a look [here](https://github.com/drakkan/sftpgo/blob/main/dataprovider/admin.go#L216)

No need to update sftpgo_api_cli, it is deprecated and it only works with v1.2.x and before.

rashmimaharana commented:
Thanks @drakkan 
Doing those changes and will update soon.

rashmimaharana commented:
Hi @drakkan 

I have done the changes mentioned. Please review once.

Thanks,
Rashmi Maharana

drakkan commented:
Hi,

you have to create a new migration, not editing an existing one, otherwise users already at migration 12 will have issues (many users use the development branch to test/use new features). 

I think a user should be allowed to edit its own email, maybe we should rename the "Credentials" page to "Profile" or something similar, however please complete this PR first. Thanks

rashmimaharana commented:
Thanks @drakkan 
Got you point. Doing those changes.

rashmimaharana commented:
Hi @drakkan 

Its done. Please review once.

Thanks,
Rashmi Maharana

drakkan commented on httpd/httpd_test.go:
```diff
@@ -11090,6 +11092,7 @@ func TestWebUserUpdateMock(t *testing.T) {
 	user.AdditionalInfo = "new additional info"
 	form := make(url.Values)
 	form.Set("username", user.Username)
+	form.Set("email", user.Email)
```
please add a check to verify that the email was really set, and do the same also if the email is set using the REST API

drakkan commented:
it looks better, thanks. Let's see if CI passes

https://github.com/drakkan/sftpgo/actions/runs/1269832460

drakkan commented:
There are a lot of warnings like this one:

```
2021-09-24T12:57:05.315 WRN database version 13 is newer than the supported one: 12
```

no time to investigate now, but probably something is wrong

rashmimaharana commented:
2021-09-24T12:57:05.315 WRN database version 13 is newer than the supported one: 12

This because i have added value 13 in DB upgrade function and its updating schema_version table with value as 13.
But not sure why its falling under "default" block of switch and giving the above warning.

coming to MySQl, i had missed the closing bracket in users table by mistake, so might be that's why it was failling. I have added  that missing closing bracket.

![image](https://user-images.githubusercontent.com/10846216/134691307-1bf2dd8b-b34f-495e-ac72-d4890a4add1b.png)
![image](https://user-images.githubusercontent.com/10846216/134691540-e4c87507-7b71-4586-ba7d-7e81687f55d8.png)

drakkan commented:
Try to set database version to 13 [here](https://github.com/drakkan/sftpgo/blob/main/dataprovider/sqlcommon.go#L22)

rashmimaharana commented:
@drakkan 
Done , updated to 13. Thanks.

drakkan commented:
Thank you. I think if you add some test cases to check the email field they will fail with the memory provider. 

I can fix this myself. If I have time this week end, I'll try to merge your patch with some minor modifications. Otherwise I'll work on this patch the next weekend.

The email field will be used in future updates for password recovery, to send data retention deletion reports or things like this.

I don't like the table you added to the UI, I'll try to rework that page to make the email field editable (maybe if the admin grants a specific permission).

rashmimaharana commented:
thanks @drakkan 
i will be happy to add changes. Please let me know, will be happy to do.

drakkan commented:
it is enough for now, thanks. I will be working on data retention this weekend, and will try to review this PR if there is some time left, possibly I will ask you to change something after a more thorough review, thank you very much appreciated

drakkan commented:
Hi,

I merged this patch after fixing some minor issues and removing the change for the credential page. I'll try to refator that page the next week. 

Obviously you remain the author of the commit

https://github.com/drakkan/sftpgo/commit/0661876e99f4691084e2175985a602d41ff771ce

thank you for your contribution, very appreciated

rashmimaharana commented:
Thanks @drakkan for all your help and direction.

drakkan commented:
thanks to you for the contribution, I'm working on a new profile page, it will look like this

![Schermata del 2021-09-28 10-53-15](https://user-images.githubusercontent.com/553263/135055961-301dc167-72d4-4479-a501-56b0907bf6a2.png)
